apps/hash-graph/lib/graph/src/api/rest.rs

```diff
@@ -223,7 +223,79 @@ async fn serve_static_schema(Path(path): Path<String>) -> Result<Response, Statu
         )
     ),
 )]
-struct OpenApiDocumentation;
+pub struct OpenApiDocumentation;
+
+impl OpenApiDocumentation {
+    /// Writes the `OpenAPI` specification to the given path.
+    ///
+    /// The path must be a directory, and the `OpenAPI` specification will be written to
+    /// `openapi.json` in that directory.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the path is not a directory, or if the files cannot be
+    /// written.
+    pub fn write_openapi(path: impl AsRef<std::path::Path>) -> Result<(), Report<io::Error>> {
+        let openapi = Self::openapi();
+        let path = path.as_ref();
+        fs::create_dir_all(path)
+            .into_report()
+            .attach_printable_lazy(|| path.display().to_string())?;
+
+        let openapi_json_path = path.join("openapi.json");
+        serde_json::to_writer_pretty(
+            io::BufWriter::new(
+                fs::File::create(&openapi_json_path)
+                    .into_report()
+                    .attach_printable("could not write openapi.json")
+                    .attach_printable_lazy(|| openapi_json_path.display().to_string())?,
+            ),
+            &openapi,
+        )
+        .map_err(io::Error::from)
+        .into_report()?;
+
+        if let Some(components) = openapi.components {
+            let schema_path_dir = path.join("components").join("schemas");
+            fs::create_dir_all(&schema_path_dir)
+                .into_report()
+                .attach_printable("could not create directory")
+                .attach_printable_lazy(|| schema_path_dir.display().to_string())?;
+
+            for (schema_path, schema) in components.schemas {
+                let schema_path = schema_path_dir.join(&schema_path);
+                let schema = serde_json::to_string_pretty(&schema)
+                    .map_err(io::Error::from)
+                    .into_report()?;
+                fs::write(schema_path, schema)?;
+            }
+        }
```

Comment:
I think we skip over this writing step as it should all be contained in the top level schema